### PR TITLE
*: add serving HTTP backup request to restore operator

### PR DESCRIPTION
**Pull request by fanminshi**
enables restore operator to listen on port 19999 for serving HTTP backup request.
the request must be in form of endpoint:19999/v1/backup/etcd_cluster_name>/?etcdVersion=<etcd_version>

### CODE REVIEW

File: pkg/controller/restore-operator/backup_handler.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
+// Copyright 2017 The etcd-operator Authors
```
Comment (hongchaodeng) [inline pkg/controller/restore-operator/backup_handler.go]: file name -> "http_handler.go" ?

**Comment by xiang90** [comment]
cluster name should be part of  the resource path. it is not a query parameter. 

the path should look like /backups/clusterName/xxx

File: pkg/controller/restore-operator/backup_handler.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+// close removes aws credential/config dir and backupDir dir which were
+// created for BackupServer.
+func (bsc *backupServerCloser) close() {
+	os.RemoveAll(bsc.awsDir)
+	os.RemoveAll(bsc.backupDir)
+}
+
+func (r *Restore) handleBackup(spec *api.BackupSpec) error {
```
Comment (hongchaodeng) [inline pkg/controller/restore-operator/backup_handler.go]: handleBackup  -> handleRestore?

File: pkg/controller/restore-operator/backup_handler.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+	"github.com/coreos/etcd-operator/pkg/backup/backend"
+	"github.com/coreos/etcd-operator/pkg/backup/backupapi"
+	backupS3 "github.com/coreos/etcd-operator/pkg/backup/s3"
+	"github.com/coreos/etcd-operator/pkg/cluster/backupstorage"
+	"github.com/sirupsen/logrus"
+)
+
+// backupServerCloser wraps around backup.BackupServer with a close method
+// that cleans up resources associated with backup.BackupServer.
+type backupServerCloser struct {
```
Comment (xiang90) [inline pkg/controller/restore-operator/backup_handler.go]: just backupServer. closer has no meaning here. no one will use it externally as an interface.

File: pkg/controller/restore-operator/operator.go [hunk 4]
```diff
@@ -59,6 +67,38 @@ func New() *Restore {
...
+	clusterName := req.URL.Query().Get("cluster")
+	if len(clusterName) == 0 {
+		http.Error(w, "cluster is not specified", http.StatusBadRequest)
+		return
+	}
+	r.mu.RLock()
+	bs, ok := r.backupServers[clusterName]
+	r.mu.RUnlock()
+	if !ok {
+		http.Error(w, fmt.Sprintf("cluster %v backup server not found", clusterName), http.StatusInternalServerError)
```
Comment (xiang90) [inline pkg/controller/restore-operator/operator.go]: return here

File: pkg/controller/restore-operator/operator.go [hunk 5]
```diff
@@ -42,6 +46,10 @@ type Restore struct {
 	kubecli       kubernetes.Interface
 	restoreCRCli  versioned.Interface
 	kubeExtClient apiextensionsclient.Interface
+
+	// mu protects backupServers.
+	mu            sync.RWMutex
+	backupServers map[string]*backupServerCloser
```
Comment (hongchaodeng) [inline pkg/controller/restore-operator/operator.go]: use concurrent map:
https://golang.org/doc/go1.9#sync-map

File: pkg/controller/restore-operator/sync.go [hunk 6]
```diff
@@ -28,6 +24,7 @@ const (
 	//
 	// 5ms, 10ms, 20ms, 40ms, 80ms, 160ms, 320ms, 640ms, 1.3s, 2.6s, 5.1s, 10.2s, 20.4s, 41s, 82s
 	maxRetries = 15
+	tmpdir     = "/tmp"
```
Comment (hongchaodeng) [inline pkg/controller/restore-operator/sync.go]: What's this var for?
Isn't it duplicate?

**Comment by fanminshi** [comment]
@xiang90 I am thinking about calling the backup http api like 
`curl -o <etcd_snap_name>  restore_operator_endpoint:19999/backup?cluster=example-cluster,etcdVersion=3.1.8`.
 I am not sure you meant by 
```
cluster name should be part of the resource path.
the path should look like /backups/clusterName/xxx
```
could you elaborate?

File: pkg/controller/restore-operator/backup_handler.go [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+		logrus.Fatalf("unknown StorageType: %v", spec.StorageType)
+	}
+	r.mu.Lock()
+	r.backupServers[spec.ClusterName] = bsc
+	r.mu.Unlock()
+	return nil
+}
+
+func createS3BackupServer(kubecli kubernetes.Interface, s3 *api.S3Source, namespace, clusterName string) (*backupServerCloser, error) {
+	awsDir, err := ioutil.TempDir(tmpdir, "")
```
Comment (hongchaodeng) [inline pkg/controller/restore-operator/backup_handler.go]: Duplicate to backup-operator/s3_backup.go ?
WIP Anyway. Will look again at final.

File: pkg/controller/restore-operator/sync.go (hunk 6 above)
Comment (fanminshi) [reply pkg/controller/restore-operator/sync.go]: yeah, it is duplicated from backup-operator.

**Comment by hongchaodeng** [comment]
> cluster name should be part of the resource path.
the path should look like /backups/clusterName/xxx

@fanminshi Just use endpoint:
```
/backups/<cluster-name>
```
For example, if cluster-name is "a", then it will query
```
/backups/a
```

File: pkg/controller/restore-operator/sync.go (hunk 6 above)
Comment (hongchaodeng) [reply pkg/controller/restore-operator/sync.go]: I mean duplicate in restore-operator/backup_handler.go too

**Comment by hongchaodeng** [comment]
I have written a prototype for the backup server in restore operator that does the cluster name path parsing. Please base on that:

branch: https://github.com/coreos/etcd-operator/tree/res_h

commit: https://github.com/coreos/etcd-operator/commit/bd3a6e9fadf62d42eef35225f506a34b126911b6

File: pkg/controller/restore-operator/s3_backup.go [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+// that cleans up resources associated with backup.BackupServer.
+type backupServer struct {
+	*backup.BackupServer
+	awsDir    string
+	backupDir string
+}
+
+// close removes aws credential/config dir and backupDir dir which were
+// created for BackupServer.
+func (bsc *backupServer) close() {
```
Comment (xiang90) [inline pkg/controller/restore-operator/s3_backup.go]: probably just remove.

Comment (fanminshi) [reply pkg/controller/restore-operator/s3_backup.go]: the method is called `close` to indicate that this `backupServer` is finished and all its related resources are released.

Comment (hongchaodeng) [reply pkg/controller/restore-operator/s3_backup.go]: I see..
@xiang90 
So the workflow here is that:
- when a Restore spec is created, it creates S3 client
- When finished serving backup, it destroys everything

The above two is concurrent, although they have implicit ordering.

I am fine with what it is now.

File: pkg/controller/restore-operator/http.go [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+	if len(clusterName) == 0 {
+		http.Error(w, "cluster is not specified", http.StatusBadRequest)
+		return
+	}
+	v, ok := r.backupServers.Load(clusterName)
+	if !ok {
+		http.Error(w, fmt.Sprintf("cluster %v backup server not found", clusterName), http.StatusInternalServerError)
+		return
+	}
+	go func() {
```
Comment (hongchaodeng) [inline pkg/controller/restore-operator/http.go]: Don't create go routine here?

File: pkg/controller/restore-operator/s3_backup.go (hunk 8 above)
Comment (hongchaodeng) [reply pkg/controller/restore-operator/s3_backup.go]: Actually after a second thought we shouldn't destroy it. A client might fail before and retry again.
We should keep it there for now. Let's handle cleanup later.

Comment (hongchaodeng) [reply pkg/controller/restore-operator/s3_backup.go]: After offline discussion, we will save a map of existing Restore CR. Restore CR will be cleaned up once user deletes the CR.

On serveBackup(), we will pull the secret, create storage backend, pull backup, serve backup, and destroy. This eliminates concurrency issue, but might have performance issue. A tradeoff for simplicity now.

**Comment by fanminshi** [comment]
cc/ @hongchaodeng I implemented the following

"On serveBackup(), we will pull the secret, create storage backend, pull backup, serve backup, and destroy. This eliminates concurrency issue, but might have performance issue. A tradeoff for simplicity now."

File: pkg/controller/restore-operator/http.go [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+)
+
+func (r *Restore) startHTTP() {
+	http.HandleFunc(backupapi.APIV1+"/backup", r.serveBackup)
+	logrus.Infof("listening on %v", listenAddr)
+	panic(http.ListenAndServe(listenAddr, nil))
+}
+
+// path: /backup/<cluster-name>
+func (r *Restore) serveBackup(w http.ResponseWriter, req *http.Request) {
```
Comment (hongchaodeng) [inline pkg/controller/restore-operator/http.go]: doc about what this method does?

File: pkg/controller/restore-operator/http.go (hunk 9 above)
Comment (hongchaodeng) [reply pkg/controller/restore-operator/http.go]: address this comment?

Comment (fanminshi) [reply pkg/controller/restore-operator/http.go]: create go routine here can speed up multiple incoming requests.

File: pkg/controller/restore-operator/http.go [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+		return
+	}
+	v, ok := r.restoreCRs.Load(clusterName)
+	if !ok {
+		http.Error(w, fmt.Sprintf("cluster %v backup server not found", clusterName), http.StatusInternalServerError)
+		return
+	}
+	// a new go routine is created to handle each request to speed up
+	// multi requests handling.
+	go func() {
```
Comment (hongchaodeng) [inline pkg/controller/restore-operator/http.go]: Again, don't create a goroutine here.
The performance concern is unnecessary for now.

Comment (fanminshi) [reply pkg/controller/restore-operator/http.go]: done.

**Comment by hongchaodeng** [comment]
lgtm

**Comment by fanminshi** [comment]
manual test:

create restore-operator:
```
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: etcd-restore-operator
spec:
  replicas: 1
  template:
    metadata:
      labels:
        name: etcd-restore-operator
    spec:
      containers:
      - name: etcd-restore-operator
        image: gcr.io/coreos-k8s-scale-testing/etcd-restore-operator:fanmin
        imagePullPolicy: Always
        command: 
        - /usr/local/bin/etcd-restore-operator
        env:
        - name: MY_POD_NAMESPACE
          valueFrom:
            fieldRef:
              fieldPath: metadata.namespace
        - name: MY_POD_NAME
          valueFrom:
            fieldRef:
              fieldPath: metadata.name
```

create restore CR:
```yaml
apiVersion: "etcd.database.coreos.com/v1beta2"
kind: "EtcdBackup"
metadata:
  name: s3-cluster-backup
spec:
  clusterName: s3-cluster-backup
  storageType: S3
  s3:
    s3Bucket: jenkins-etcd-operator
    prefix: prefix
    awsSecret: aws
```

get snap:
```sh
curl -o snap http://172.17.0.5:19999/v1/backup/s3-cluster-backup?etcdVersion=3.1.8
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 24608    0 24608    0     0   123k      0 --:--:-- --:--:-- --:--:--  123k
```

**Comment by fanminshi** [comment]
merge when green.

**Pull request merged**